src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+                {
+                    ItemStack stack = workerInventory.getStackInSlot(i);
+                    if(stack == null)
+                    {
+                        continue;
+                    }
+
+                    if(InventoryUtils.addItemStackToInventory(buildingToDeliver.getTileEntity(), stack))
+                    {
+                        workerInventory.removeStackFromSlot(i);
+                    }
+                }
+                buildingToDeliver.setOnGoingDelivery(false);
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Prepare deliveryman for delivery.
+     * Check if the building still needs the item and if the required items are still in the warehouse.
+     * @return the next state to go to.
+     */
+    private AIState prepareDelivery()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if(ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if(buildingToDeliver != null)
+            {
+                boolean ableToDeliver = wareHouse.getTileEntity().checkInWareHouse(buildingToDeliver);
+                if(!ableToDeliver)
+                {
+                    return START_WORKING;
+                }
+                itemsToDeliver = new ArrayList<>(buildingToDeliver.getNeededItems());
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    private AIState gatherItemsFromWareHouse()
+    {
+        AbstractBuildingWorker ownBuilding = getOwnBuilding();
+        if (ownBuilding instanceof BuildingDeliveryman)
+        {
+            AbstractBuilding buildingToDeliver = ((BuildingDeliveryman) ownBuilding).getBuildingToDeliver();
+            if (buildingToDeliver != null)
+            {
+                if(itemsToDeliver.isEmpty() && hasTools(buildingToDeliver))
+                {
+                    return DELIVERY;
+                }
+
+                gatherItems(buildingToDeliver);
+                return GATHER_IN_WAREHOUSE;
+            }
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Check if the deliveryman has all the tools to make the delivery.
+     * @param buildingToDeliver the building to deliver to.
+     * @return true if is ready to deliver.
+     */
+    private boolean hasTools(final AbstractBuilding buildingToDeliver)
+    {
+        String requiredTool = buildingToDeliver.getRequiredTool();
+        if(requiredTool.isEmpty())
+        {
+            return true;
+        }
+
+        if(requiredTool.equals(Utils.PICKAXE))
+        {
+            return InventoryUtils.isPickaxeInTileEntity(worker.getInventoryCitizen(), buildingToDeliver.getNeededPickaxeLevel());
+        }
+        return InventoryUtils.isToolInTileEntity(worker.getInventoryCitizen(), requiredTool);
+    }
+
+    /**
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) First sentence should end with a period. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=checkstyle%3AJavadoc_Style)